### PR TITLE
#72 Now is possibile to pass IRenderFunction to component Render*

### PR DESCRIPTION
**Pull request by MarcOne68**
I defined the class RenderOverride. 
You can create an instance of this class passing a IRenderFunction<...> and set the Render* property of a component with that instance.

### CODE REVIEW

File: libs/core/src/lib/components/wrapper-component.ts [hunk 1]
```diff
@@ -213,6 +219,10 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
       return (context: TContext) => htmlRenderer.render(context);
     }
 
+    if (input instanceof RenderOverride) {
+      return undefined;
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: why not `return input.renderer`?

Comment (MarcOne68) [reply libs/core/src/lib/components/wrapper-component.ts]: How I wrote in Issue #72, these changes help me to set the styles 
This is the code example:

```
private overrideRenderRow(props: IDetailsRowProps, defaultRender: (props: IDetailsRowProps) => JSX.Element): JSX.Element {
  props.styles = MY_STYLES;
  return defaultRender(props);
}
```

File: libs/core/src/lib/components/wrapper-component.ts [hunk 2]
```diff
@@ -28,6 +28,7 @@ import { createComponentRenderer, createHtmlRenderer, createTemplateRenderer } f
 import { toObject } from '../utils/object/to-object';
 import { afterRenderFinished } from '../utils/render/render-delay';
 import { unreachable } from '../utils/types/unreachable';
+import { IRenderFunction } from 'office-ui-fabric-react';
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Please don’t import stuff from `office-ui-fabric-react` inside the `@angular-react/core` package - we want to keep it library-agnostic.

A more generic type is needed here instead therefore.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 1 above)
Comment (bengry) [reply libs/core/src/lib/components/wrapper-component.ts]: Sounds reasonable - you want to expose the native JSX to Angular-land then. This should be fine for simple cases, since JSX in Angular is something I’d like to avoid (although possible of course).

However, I suggest we change the API, to avoid the boilerplate of creating a class (which would really only have one instance), if possible.
I want to think about an API further, and see if we can somehow avoid exposing JSX to the final user (i.e. the developer using `@angular-react/fabric`) - perhaps allowing to pass a function that accepts `TProps` and returns `TProps`, which will then be fed to the `defaultRenderer` that `office-ui-fabric-react` exposes. If you have an idea for an API - I’m open for suggestions, and how it would work.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 3]
```diff
@@ -219,17 +218,17 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
...
-    }
-
     if (typeof input === 'object') {
-      const { componentType, factoryResolver, injector } = input;
-      const componentFactory = factoryResolver.resolveComponentFactory(componentType);
-      const componentRef = componentFactory.create(injector);
-
-      // Call the function again with the created ComponentRef<TContext>
-      return this.createInputJsxRenderer(componentRef, additionalProps);
+      if ('componentType' in input) {
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: please replace this with a [user defined type guard](https://basarat.gitbooks.io/typescript/docs/types/typeGuard.html#user-defined-type-guards), this function does too much at the moment (it was bearable so far since it only had to deal with one `object`, but in hindsight it should've extracted the "is this a `RenderComponentOptions`" to a user defined type guard.

Also, you want to check with `hasOwnProperty`, since it doesn't look up the prototype chain.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 4]
```diff
@@ -219,17 +218,17 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
...
+        const { componentType, factoryResolver, injector } = input;
+        const componentFactory = factoryResolver.resolveComponentFactory(componentType);
+        const componentRef = componentFactory.create(injector);
+
+        // Call the function again with the created ComponentRef<TContext>
+        return this.createInputJsxRenderer(componentRef, additionalProps);
+      } else {
+        return undefined;
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: why is this `undefined` in the `DefaultRenderOptions` case?

File: libs/core/src/lib/components/wrapper-component.ts [hunk 5]
```diff
@@ -248,8 +247,10 @@ export abstract class ReactWrapperComponent<TProps extends {}> implements AfterV
       additionalProps?: ReactContentProps;
     }
   ): (props?: TProps, defaultRender?: JsxRenderFunc<TProps>) => JSX.Element | null {
-    if (renderInputValue instanceof RenderOverride) {
-      return renderInputValue.renderer;
+    if ((typeof renderInputValue === 'object') && ('getProps' in renderInputValue)) {
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Please use a user-defined type guard here, as mentioned in the other comment. Making one will also make sure you can re-use the logic of it in multiple places.

File: libs/core/src/lib/components/wrapper-component.ts [hunk 6]
```diff
@@ -41,16 +40,16 @@ export interface RenderComponentOptions<TContext extends object> {
   readonly injector: Injector;
 }
 
-export class RenderOverride<TContext extends object> {
-  constructor(public renderer: IRenderFunction<TContext>) { }
+export interface DefaultRenderOptions<TContext extends object> {
```
Comment (bengry) [inline libs/core/src/lib/components/wrapper-component.ts]: Thinking out load here - the `DefaultRenderOptions` doesn't convey the message that it allows mutating the props. If you have suggestions for a better/other name - I'm open for suggestions, but if not - please add a JDoc comment explaining what the purpose of this object is.

File: libs/core/src/lib/components/wrapper-component.ts (hunk 4 above)
Comment (bengry) [reply libs/core/src/lib/components/wrapper-component.ts]: I understand the technical reason of. Why needing to handle this case is here - but why is returning undefined specifically? Shouldn't there be a render function call here, or rather, a function that creates a render function.

Comment (MarcOne68) [reply libs/core/src/lib/components/wrapper-component.ts]: Only because `unreachable(input)` throw an error during the compilation.
The execution never reach that line of code because the method `createRenderPropHandler` executes the cases RenderPropOptions and RenderPropContext.

File: libs/fabric/src/lib/components/text-field/base-text-field.component.ts [hunk 7]
```diff
@@ -68,6 +68,7 @@ export class FabBaseTextFieldComponent extends ReactWrapperComponent<ITextFieldP
   @Output() readonly onChange = new EventEmitter<{ event: Event; newValue?: string }>();
   @Output() readonly onBeforeChange = new EventEmitter<{ newValue: any }>();
   @Output() readonly onNotifyValidationResult = new EventEmitter<{ errorMessage: string; value: string | undefined }>();
+  @Output() readonly onBlur = new EventEmitter<{ event: Event }>();
```
Comment (bengry) [inline libs/fabric/src/lib/components/text-field/base-text-field.component.ts]: Please move these changes to a separate PR. It makes it easier to maintain and create the changelog (this part of the PR looks fine as it is, so if you move it it can get it as-is)

Comment (MarcOne68) [reply libs/fabric/src/lib/components/text-field/base-text-field.component.ts]: OK, I'll try to do it but.... I don't know so well GIT :)

Comment (bengry) [reply libs/fabric/src/lib/components/text-field/base-text-field.component.ts]: Basically create a new branch from `master` and make the same changes there (or you can cherry-pick the commit using `git cherry-pick 95e1991` when you're on the new branch.
Then create a new PR from it.

**Comment by bengry** [comment]
@MarcOne68 This is a good PR (or two, as I mentioned in another comment).
Is this something you're still working on and would like to merge? It's been over a month since the last change here.

**Comment by bengry** [comment]
Implemented as part of #92 and #99.
